### ISSUE TITLE
False positive: Probable bugs -> Array and string offset validity

### ISSUE THREAD

**Issue by tfettig01** [kalessil/phpinspectionsea#809]
Caused by using https://redis.io/commands/get,  json_encode, and casting the results to an array.

https://github.com/phpredis/phpredis#get
http://php.net/manual/en/function.json-encode.php

This is some simplified emulation code that's showing the false positive:
```php
<?php

/**
 * @return string|bool
 */
function foo()
{
    return rand(0, 1) ? json_encode('bar') : false;
}

$bar = foo();
$bar = (array)json_decode($bar, true);

// false positive on next line
echo isset($bar[0]) ? 'yes' : 'no';
```

https://3v4l.org/IOupG

**Comment by kalessil** [comment]
Thank you for reporting @tfettig01, I'll check what can we do here.

**Comment by kalessil** [comment]
@tfettig01 : seems this needs to be reported to PhpStorm team as following works properly:

```php
/** @return string|bool */
function foo() {
    return random_int(0, 1) ? json_encode('bar') : false;
}

$bar = (array)json_decode(foo(), true);
echo isset($bar[0]) ? 'yes' : 'no';
```
To be more specific: PhpStorm merges types instead of overriding them in case of sequential assignments. That's quite unexpected behaviou.

**Comment by tfettig01** [comment]
Thank you.

I reported it at https://youtrack.jetbrains.com/issue/WI-41106 and tagged you in a comment.  Hopefully you don't mind.

**Comment by kalessil** [comment]
Thank you for reporting. Yes, I'm totally fine with it =)
